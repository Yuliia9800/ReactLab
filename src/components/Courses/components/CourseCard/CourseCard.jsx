import React from 'react';

import { Button } from 'common';
import { pipeDuration, dateGenerator, getAuthorName } from 'helpers';
import { SHOW_COURSE_BUTTON_TEXT } from 'constants/constants';

function CourseCard({ title, description, authors, duration, creationDate }) {
	const authorsNames = authors.map((author) => getAuthorName(author));

	return (
		<div className='card flex min-h-[200px] flex-row justify-between gap-10 bg-base-200 p-5 shadow-xl'>
			<div className='w-4/6 space-y-2 '>
				<h1 className='text-2xl font-bold'>{title}</h1>
				<p className='text-sm'>{description}</p>
			</div>

			<div className='flex w-2/6 flex-col justify-between space-y-2'>
				<div className=' space-y-2'>
					<p className='truncate text-sm'>
						<b>Author: </b>
						{authorsNames.join(', ')}
					</p>
					<p className='text-sm'>
						<b>Duration: </b>
						{pipeDuration(duration)} hours
					</p>
					<p className='text-sm'>
						<b>Created: </b>
						{dateGenerator(creationDate)}
					</p>
				</div>
				<Button
					buttonText={SHOW_COURSE_BUTTON_TEXT}
					onClick={() => console.log(SHOW_COURSE_BUTTON_TEXT)}
					className='btn-default btn-block self-center'
				/>
			</div>
		</div>
	);
}

export default CourseCard;
