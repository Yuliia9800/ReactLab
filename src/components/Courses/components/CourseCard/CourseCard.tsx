import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { BsTrash, BsPencil } from 'react-icons/bs';

import { Button } from 'common';
import { pipeDuration, dateGenerator } from 'helpers';
import { Course } from 'types';
import { SHOW_COURSE_BUTTON_TEXT } from 'constant';
import { deleteCourse } from 'store/courses';
import { AppDispatch, RootState } from 'store';

interface CourseCardProps extends Course {}

function CourseCard({
	id,
	title,
	description,
	authors,
	duration,
	creationDate,
}: CourseCardProps) {
	const navigate = useNavigate();
	const dispatch = useDispatch<AppDispatch>();
	const authorsList = useSelector((state: RootState) => state.authors);
	const role = useSelector((state: RootState) => state.user.role);

	const handleShowCourse = () => {
		navigate(`/courses/${id}`);
	};

	const courseAuthors = authors.map(
		(authorId) => authorsList?.find(({ id }) => id === authorId)?.name
	);

	const handleEdit = () => {
		navigate(`/courses/update/${id}`);
	};

	const handleDelete = () => {
		dispatch(deleteCourse(id));
	};

	return (
		<div className='card flex min-h-[200px] flex-row justify-between gap-10 bg-base-200 p-5 shadow-xl'>
			<div className='w-4/6 space-y-2'>
				<h1 className='text-2xl font-bold'>{title}</h1>
				<p className='text-sm'>{description}</p>
			</div>

			<div className='flex w-2/6 min-w-[300px] flex-col justify-between space-y-2'>
				<div className='space-y-2'>
					<p className='truncate text-sm'>
						<b>Author: </b>
						{courseAuthors.join(', ')}
					</p>
					<p className='text-sm'>
						<b>Duration: </b>
						{pipeDuration(duration)} hours
					</p>
					<p className='text-sm'>
						<b>Created: </b>
						{dateGenerator(creationDate)}
					</p>
				</div>
				<div className='flex flex-wrap justify-between '>
					<Button
						buttonText={SHOW_COURSE_BUTTON_TEXT}
						onClick={handleShowCourse}
						className=''
					/>

					{role === 'admin' ? (
						<div className='space-x-2'>
							<Button
								buttonText={<BsPencil />}
								onClick={handleEdit}
								className='text-xl'
							/>
							<Button
								buttonText={<BsTrash />}
								onClick={handleDelete}
								className='text-xl '
							/>
						</div>
					) : null}
				</div>
			</div>
		</div>
	);
}

export default CourseCard;
