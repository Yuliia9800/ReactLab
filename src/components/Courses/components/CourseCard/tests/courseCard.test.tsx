import React from 'react';

import { renderWithProviders, screen } from 'testing';
import CourseCard from '../CourseCard';

test('should display correct props', () => {
	const props = {
		id: '1',
		title: 'title',
		description: 'description',
		authors: ['11'],
		duration: 70,
		creationDate: '10-10-2022',
	};

	const state = {
		authors: [
			{
				id: '11',
				name: 'author11',
			},
		],
	};

	renderWithProviders(<CourseCard {...props} />, { state });

	expect(screen.getByText(props.title)).toBeInTheDocument();
	expect(screen.getByText(props.description)).toBeInTheDocument();
	expect(screen.getByText('01:10 hours')).toBeInTheDocument();
	expect(screen.getByText('10.10.2022')).toBeInTheDocument();
	expect(screen.getByText('author11')).toBeInTheDocument();
});
