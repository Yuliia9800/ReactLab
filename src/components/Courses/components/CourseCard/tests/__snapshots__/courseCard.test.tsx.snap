// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="card flex min-h-[200px] flex-row justify-between gap-10 bg-base-200 p-5 shadow-xl"
    data-testid="course"
  >
    <div
      class="w-4/6 space-y-2"
    >
      <h1
        class="text-2xl font-bold"
      >
        title
      </h1>
      <p
        class="text-sm"
      >
        description
      </p>
    </div>
    <div
      class="flex w-2/6 min-w-[300px] flex-col justify-between space-y-2"
    >
      <div
        class="space-y-2"
      >
        <p
          class="truncate text-sm"
        >
          <b>
            Author: 
          </b>
        </p>
        <p
          class="text-sm"
        >
          <b>
            Duration: 
          </b>
          01:10
           hours
        </p>
        <p
          class="text-sm"
        >
          <b>
            Created: 
          </b>
          10.10.2022
        </p>
      </div>
      <div
        class="flex flex-wrap justify-between "
      >
        <button
          class="btn-sm btn px-5"
          type="button"
        >
          Show Course
        </button>
      </div>
    </div>
  </div>
</div>
`;
